lib/ash_authentication/jwt/config.ex

```diff
@@ -77,6 +82,51 @@ defmodule AshAuthentication.Jwt.Config do
     "~> #{vsn.major}.#{vsn.minor}"
   end
 
+  @doc """
+  The generator function used to generate the "tenant" claim.
+
+  If the token resource or the user resource have tenancy enabled then we
+  validate the presence of the `tenant` option.
+  """
+  @spec generate_tenant(Resource.t(), Keyword.t()) :: nil | String.t()
+  def generate_tenant(resource, options) do
+    if multitenancy_required?(resource) do
+      case Keyword.fetch(options, :tenant) do
+        {:ok, tenant} -> tenant
+        :error -> raise TenantRequired, resource: resource
+      end
+    end
+  end
+
+  @doc """
+  Validate that the "tenant" claim matches the provided tenant option.
+
+  If the token resource or the user resource have tenancy enabled then we
+  validate the value of the "token" claim.
+  """
+  @spec validate_tenant(String.t(), Resource.t(), Keyword.t()) :: boolean()
+  def validate_tenant(tenant, resource, options) do
+    if multitenancy_required?(resource) do
+      case Keyword.fetch(options, :tenant) do
+        {:ok, ^tenant} -> true
+        _ -> false
+      end
+    else
+      true
+    end
+  end
+
+  defp multitenancy_required?(user_resource) do
```

Comment:
It's not about if the token is multi tenant, but if the user is being multi tenant. When we generate the token we need to set the tenant into it, but only if the tenant option is set when generating the token. You can have resources that support being accessed both with a tenant and without.

Reply:
Actually I think we shouldn't worry about any of it. If the tenant option is set (to anything other than nil) in this code, then we write the tenant into the claims. If there is a tenant claim when validating, or if the tenant option is set, then we validate that the two match.

Reply:
Nope, I'm wrong. We do want to check if the user (but not the token) supports multitenancy, and if so then we add and validate the tenant. That might make it easier because in *those* cases it actually makes sense to add a nil tenant claim if it is nil.

Reply:
Actually I think I just realised that we can just make it optional whether we add it to the config if required.